File: 3524-find-x-value-of-array-i/3524-find-x-value-of-array-i.java
class Solution {
    public long[] resultArray(int[] nums, int k) {
        long[] ans = new long[k];
        long[] dp = new long[k];

        for (int num : nums) {
            int x = num % k;
            long[] next = new long[k];
            next[x]++;

            for (int r = 0; r < k; r++) {
                int newR = (r * x) % k;
                next[newR] += dp[r];
            }

            for (int r = 0; r < k; r++) {
                ans[r] += next[r];
            }
            dp = next;
        }
        return ans;
    }
}

File: 3524-find-x-value-of-array-i/3524-find-x-value-of-array-i.py
class Solution:
    def resultArray(self, nums: List[int], k: int) -> List[int]:
        ans = [0] * k
        dp = [0] * k

        for num in nums:
            x = num % k
            nxt = [0] * k
            nxt[x] += 1

            for r in range(k):
                new_r = (r * x) % k
                nxt[new_r] += dp[r]

            for r in range(k):
                ans[r] += nxt[r]
            dp = nxt

        return ans